### PR TITLE
fix(daemon): ignore setlocal in Windows task parsing

### PR DESCRIPTION
**Pull request by ly85206559**
## Summary
- Problem: `openclaw gateway status` on Windows could flag a healthy Scheduled Task as missing the `gateway` subcommand when the generated task script contained batch prologue directives like `setlocal` before the real command.
- Why it matters: this produced a false-positive service audit warning in `gateway status` / `doctor`, making valid Windows installs look broken.
- What changed: `readScheduledTaskCommand()` now classifies `setlocal` / `endlocal` batch directives explicitly, keeps skipping unsupported `set` forms, and now continues through same-line chained prologue statements before capturing the real gateway command.
- What did NOT change (scope boundary): this PR does not change task installation, task execution, or Windows runtime control flow outside command-line parsing.

## Change Type (select all)
- [x] Bug fix
- [ ] Feature
- [ ] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)
- [x] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR
- Closes #53474
- Related #49126
- [x] This PR fixes a bug or regression

## Root Cause (if applicable)
- Root cause: the Windows task-script scanner treated each line as a single batch statement and skipped only `set ` assignments, so `setlocal` / `endlocal` could be misread as the runnable command and same-line chained prologue directives could hide the real gateway command.
- Missing detection / guardrail: there was no regression coverage for Windows task scripts that include batch prologue directives before the real `gateway` command.
- Contributing context (if known): the installed/inspected script format is batch syntax, but the parser only handled the simplest one-statement-per-line subset.

## Regression Test Plan (if applicable)
- Coverage level that should have caught this:
  - [x] Unit test
  - [ ] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file: `src/daemon/schtasks.test.ts`
- Scenario the test should lock in: ignore `setlocal` / `endlocal`, ignore unsupported `set` forms, and continue through `&` / `&&` / `||` chained batch prologue statements until the real gateway command is found.
- Why this is the smallest reliable guardrail: the bug is isolated to `readScheduledTaskCommand()` batch parsing, so focused unit coverage exercises the exact failing path without needing real Windows task registration.
- Existing test that already covers this (if any): None.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes
- `openclaw gateway status` / `openclaw doctor` no longer report a false-positive Windows service audit warning when the inspected task script uses `setlocal` / `endlocal` prologue directives before the real gateway command.

## Diagram (if applicable)
```text
Before:
[gateway status] -> [scan gateway.cmd] -> [treat setlocal/prologue as command] -> [false audit warning]

After:
[gateway status] -> [scan gateway.cmd] -> [skip/unwrap batch prologue directives] -> [parse real gateway command] -> [correct audit result]
```

## Security Impact (required)
- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation: N/A

### Environment
- OS: Windows 10 (native)
- Runtime/container: Node.js 24 / Vitest daemon project
- Model/provider: N/A
- Integration/channel (if any): Windows Scheduled Task gateway runtime
- Relevant config (redacted): default Windows gateway task script under `%USERPROFILE%\\.openclaw\\gateway.cmd`

### Steps
1. Install or inspect a Windows gateway task script whose batch prologue includes `setlocal` / `endlocal` before the real gateway command.
2. Run `openclaw gateway status` or call `readScheduledTaskCommand()` through the daemon tests.
3. Observe which command line gets parsed.

### Expected
- The parser skips batch prologue directives and returns the real gateway command line.

### Actual
- Before this fix, `setlocal` could be treated as the runnable command and chained prologue directives could prevent the real gateway command from being found.

## Evidence
- [x] Failing test/log before + passing after
- [ ] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)
- Verified scenarios:
  - `pnpm test src/daemon/schtasks.test.ts`
  - `pnpm test src/daemon/schtasks.install.test.ts src/daemon/schtasks.stop.test.ts`
  - reviewed the exact issue root cause in `src/daemon/schtasks.ts` against issue #53474 and the Windows tracker entry in #49126
- Edge cases checked:
  - unsupported `set` forms like `set /a ...` and bare `set VAR`
  - bare `setlocal`
  - `endlocal`
  - same-line chained `setlocal ... && ...` and `endlocal || ...`
- What you did **not** verify:
  - `pnpm check` still stalled locally after `pnpm check:madge-import-cycles` with no further output, matching the behavior already noted on this PR

## Review Conversations
- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration
- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations
- Risk: widening parsing to same-line batch chains could accidentally split command text on command separators that are part of shell syntax rather than batch prologues.
  - Mitigation: split only on unquoted / unescaped `&`, `&&`, and `||`, keep `2>&1`-style redirections intact, and lock behavior with focused regression tests.

### CODE REVIEW

File: src/daemon/schtasks.ts [hunk 1]
```diff
@@ -108,6 +108,53 @@ function resolveTaskUser(env: GatewayServiceEnv): string | null {
...
+  if (lower.startsWith("set ")) {
+    const assignment = parseCmdSetAssignment(line.slice(4));
+    if (assignment) {
+      return {
+        kind: "env_assignment",
+        key: assignment.key,
+        value: assignment.value,
+      };
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts:238]: Reply to @greptile-apps

Thanks for the review — you’re right. When parseCmdSetAssignment returned null, the new classifier fell through to kind: "command", which regressed the old behavior where every set line was skipped until we found a real command. We fixed this by returning env_scope after the assignment branch when parsing fails, so those lines are skipped the same way as setlocal / endlocal. Added a regression test covering non-assignment forms (e.g. set /a without =, bare set VAR). Pushed in the latest commit on this branch.

File: src/daemon/schtasks.ts [hunk 2]
```diff
@@ -108,6 +108,53 @@ function resolveTaskUser(env: GatewayServiceEnv): string | null {
...
+  if (lower === "setlocal" || lower.startsWith("setlocal ")) {
+    return { kind: "env_scope" };
+  }
+  if (lower === "endlocal" || lower.startsWith("endlocal ")) {
+    return { kind: "env_scope" };
+  }
+  if (lower.startsWith("set ")) {
+    const assignment = parseCmdSetAssignment(line.slice(4));
+    if (assignment) {
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts:229]: Agreed — unsupported SET forms that don’t parse as assignments were incorrectly treated as the gateway command. We now classify them as env_scope so the scanner keeps going to the actual command line, matching the pre-refactor loop. Regression test added; fix is in the latest push.

File: src/daemon/schtasks.ts [hunk 3]
```diff
@@ -108,6 +108,55 @@ function resolveTaskUser(env: GatewayServiceEnv): string | null {
...
+    return { kind: "echo" };
+  }
+  if (lower.startsWith("rem ")) {
+    return { kind: "comment" };
+  }
+  if (lower === "setlocal" || lower.startsWith("setlocal ")) {
+    return { kind: "env_scope" };
+  }
+  if (lower === "endlocal" || lower.startsWith("endlocal ")) {
+    return { kind: "env_scope" };
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts]: Addressed in `cef5a1ab13`. The parser now walks same-line batch chains before choosing the runnable command, so `setlocal ... &/&&/|| ...` and `endlocal ... &/&&/|| ...` no longer cause `readScheduledTaskCommand()` to miss the real gateway command. Added regression coverage in `src/daemon/schtasks.test.ts` for both `setlocal && ...` and `endlocal || ...` cases.

File: src/daemon/schtasks.ts [hunk 4]
```diff
@@ -108,6 +108,112 @@ function resolveTaskUser(env: GatewayServiceEnv): string | null {
...
+  statement: string;
+  remainder: string;
+};
+
+function isBackslashEscapedQuote(value: string, index: number): boolean {
+  let backslashCount = 0;
+  for (let cursor = index - 1; cursor >= 0 && value[cursor] === "\\"; cursor -= 1) {
+    backslashCount += 1;
+  }
+  return backslashCount % 2 === 1;
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts]: Addressed in `8fd41227cf`. `splitFirstBatchStatement()` now follows the same `backslash-quote-only` rule as `parseCmdScriptCommandLine()`: if a `\"` sequence is present, the quote stays inside the current quoted argument instead of toggling quote state. That keeps later `&` / `&&` / `||` inside rendered arguments from being treated as batch separators, including the `\\\"` case produced when the original argument already contains a backslash before a quote. I also added a regression test that builds the argument via `quoteCmdScriptArg('a\\"b&c')` and confirms `--port` still parses correctly.

**Comment by ly85206559** [comment]
### Maintainer update: rebase, squash, CI notes

- **Rebased** this branch onto the latest \main\ and resolved no conflicts (clean replay of the 6 original commits).
- **Squashed** into a single commit: \Daemon: fix Windows gateway.cmd task parsing for status audit (#53474)\, and added a matching **CHANGELOG** entry under Unreleased → Fixes (\#53474\).

**CI (\check\ / \check-additional\) vs this PR**

- The daemon changes are covered by \pnpm test src/daemon/schtasks.test.ts src/daemon/schtasks.install.test.ts src/daemon/schtasks.stop.test.ts\ (46 tests) — all green locally.
- Full \pnpm check\ in this workspace failed in \	sgo:extensions\ on missing \@copilotkit/aimock\ under \extensions/qa-lab\ — **unrelated** to \src/daemon/schtasks.ts\. So the failing \check\ job on the PR run is **not treated as caused by this diff** unless CI shows a failure in daemon/lint for these files.

**Bot review**

- Greptile/Codex threads on unparseable \set\ lines, chained prologues (\&\/\&&\/\||\), escaped quotes, and single-pipe splitting were **already addressed** in the pre-squash history; the squash **preserves** that behavior. If any thread should stay open for maintainer judgment, call it out and I will follow up.

Thanks for review.

**Comment by ly85206559** [comment]
## Codex P2 (`endlocal` and env snapshot)

Addressed in `readScheduledTaskCommand`: on `setlocal` we push a shallow copy of the accumulated `environment`; on `endlocal` we pop and `restoreEnvironmentSnapshot` so `SET` assignments inside the local block do not apply to the parsed gateway command after `endlocal` (matches cmd local scope and avoids stale values for `resolveScheduledTaskPort`).

- `classifyBatchScriptLine` now returns distinct kinds for `setlocal` / `endlocal` (instead of generic `env_scope`) to drive the stack.
- Tests: updated the case where `FOO` was set inside `setlocal` and cleared after `endlocal`; added a regression for shadowed `OPENCLAW_GATEWAY_PORT` restored to the outer value.

Commit: `b26c8106df` (single amended commit).

**Comment by ly85206559** [comment]
## Codex P2 — pipe LHS vs RHS (#66062)

Addressed **“Isolate parser state when crossing single-pipe statements”**:

- `splitFirstBatchStatement` now returns `usedSinglePipe` when the separator is a single `|` (CMD pipe, not `||`).
- In `readScheduledTaskCommand`, when `usedSinglePipe` and there is a non-empty RHS, we **suppress mutations** from the LHS: no `set` / `cd /d` / `setlocal` / `endlocal` side effects on the accumulated `environment`, `workingDirectory`, or `setlocal` stack, and we do not treat a LHS-only `command` as the gateway line (the runnable command is after the pipe).

Regression tests:

- `set OPENCLAW_GATEWAY_PORT=99999 | node gateway.js …` → parsed command has **no** that env var.
- `cd /d "…" | node gateway.js …` → **no** `workingDirectory` from the pipe LHS.

Commit: `e1fc5e5157`.

**Comment by ly85206559** [comment]
## Codex P2 — preserve gateway command on LHS of `|`

Agreed: `suppressPipeStageMutations` must only block **side effects** (`set` / `cd /d` / `setlocal` / `endlocal`) on the pipe LHS, not **`command`** tokens.

- Removed the `continue` that skipped `kind: "command"` when `suppressPipeStageMutations` was true, so lines like `node gateway.js --port 18789 | findstr …` resolve to the **left-hand** launch line (the real gateway process), not the RHS filter.

- Comment updated to state that the LHS can still be the runnable gateway command.

- Test: `keeps the gateway launch on the left side of a pipe (e.g. piped to findstr)`.

Commit: `25eee6aaa2`.

File: src/daemon/schtasks.ts [hunk 5]
```diff
@@ -117,31 +249,60 @@ export async function readScheduledTaskCommand(
...
-          environment[assignment.key] = assignment.value;
+      let remainder = rawLine;
+      while (remainder.trim()) {
+        const split = splitFirstBatchStatement(remainder);
+        remainder = split.remainder;
+        // Pipe stages run in separate cmd contexts; do not let LHS `set`/`cd`/`setlocal`
+        // affect env/cwd for the RHS. The LHS may still be the real gateway command
+        // (e.g. `node gateway.js | findstr ...`); only side-effect directives are suppressed.
+        const suppressPipeStageMutations =
+          split.usedSinglePipe && split.remainder.trim() !== "";
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts]: Addressed in c48eecda75953b01ba1aec5a03c4d3dea6436189.

We now track whether a single `|` already appeared on the **current physical line**. Any later stage on that same line (`set`, `cd /d`, `setlocal`, `endlocal`) suppresses persisted mutations to `environment` / `workingDirectory` / `setlocal` stack — so RHS pipe stages cannot leak assignments into subsequent script lines while still preserving the gateway `command` from whichever stage carries it (`node … | findstr` unchanged).

Also treat a bare `rem` as `comment` (matches cmd.exe); otherwise `rem | …` could mis-parse `rem` as the gateway command.

Regression: `src/daemon/schtasks.test.ts` — “does not apply set from a right-hand pipe stage to following script lines”.

**Comment by ly85206559** [comment]
## Maintainer update: CI fix, rebase, squash

- **CI (check-lint / plugin-sdk boundary dts):** Failed with TS2678 on an unreachable `case "command"` in `readScheduledTaskCommand` after `kind === "command"` is handled earlier. Removed the dead switch arm; oxlint on `src/daemon/schtasks.ts` passes locally.

- **Rebased** onto latest `origin/main` (819d15481d).

- **Squashed** branch history into **one commit:** `e4983dfc208745bdc761de848fc2046020282545` — `fix(daemon): parse Windows gateway.cmd tasks for audit (setlocal, pipes, RHS staging) (#53474)`.

- **Regression:** `pnpm test src/daemon/schtasks.test.ts` (40 tests) and `pnpm check:changed` green locally.

Please re-run CI on the PR; failures tied to the lint/type gate should clear.

File: src/daemon/schtasks.ts [hunk 6]
```diff
@@ -108,6 +108,138 @@ function resolveTaskUser(env: GatewayServiceEnv): string | null {
...
+    // Single `|` is CMD pipe; split so prologue directives (setlocal, set, …) do not
+    // hide the real command on the same line (e.g. `setlocal … | node gateway.js`).
+    if (current === "|") {
+      return {
+        statement: rawLine.slice(0, index).trim(),
+        remainder: rawLine.slice(index + 1).trim(),
+        usedSinglePipe: true,
+      };
+    }
+    if (current === "&" && rawLine[index - 1] !== ">") {
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts]: Fixed in latest push (`5737020fb2`): `splitFirstBatchStatement` now skips unquoted single `&` when the previous character is `<`, same as `>` for `2>&1`, so `<&n` handle-duplication stays on one statement and flags like `--port` are not truncated. Regression: `does not split batch chains on \\\`<&\\\` input-handle redirection (#66062)`.

**Comment by ly85206559** [comment]
## Update (rebase + bot follow-up)

- Rebased ix/gh-53474-schtasks-setlocal onto latest main (no conflicts).
- **Codex (line ~281, pipe RHS staging + later |):** when the chosen gateway line is the LHS of a *second* single-pipe (e.g. ... && node ... | findstr), we now merge pipeRhsStaging before recording the command so set/cd in that stage are not dropped. Pushed as a single amended commit: \d71105f194\.
- **Tests:** \pnpm test src/daemon/schtasks.test.ts\ — 42 passed.

Other Greptile/Codex items in the thread were already covered in earlier history; this commit contains the full squashed history plus the second-pipe merge fix above.

File: src/daemon/schtasks.ts [hunk 7]
```diff
@@ -117,31 +254,106 @@ export async function readScheduledTaskCommand(
...
+        remainder = split.remainder;
+        const lhsPipeStage =
+          split.usedSinglePipe && split.remainder.trim() !== "";
+        const parsedLine = classifyBatchScriptLine(split.statement);
+
+        if (parsedLine.kind === "command") {
+          if (lhsPipeStage) {
+            // Runnable gateway on pipe LHS (e.g. `node … | findstr`); RHS staging N/A.
+            commandLine = parsedLine.raw;
+            break;
```
[the remark this answers is not in the document]
Comment (ly85206559) [reply src/daemon/schtasks.ts:289]: Addressed in d71105f194. readScheduledTaskCommand now merges pipeRhsStaging into the accumulated env/cwd before taking a command that is the left side of a later single | (e.g. rem | set OPENCLAW_GATEWAY_PORT=9999 && node gateway.js | findstr ...), so staged set/cd in the same pipe stage still apply. Regression: applies staged RHS env before a pipe-stage command when a second pipe follows (e.g. findstr) in src/daemon/schtasks.test.ts.

**Comment by ly85206559** [comment]
The issue is closed,so the pr will close too
